lib/puppet/util/windows/taskscheduler2.rb

```diff
@@ -0,0 +1,707 @@
+      :minute => matches['minute'],
+      :hour => matches['hour'],
+      :second => matches['second'],
+    }
+  end
+
+  def duration_hash_to_seconds(value)
+    time = 0
+    # Note - the Year and Month calculations are approximate
+    time = time + value[:year].to_i   * (365.2422 * 24 * 60**2).to_i unless value[:year].nil?
+    time = time + value[:month].to_i  * (365.2422 * 2 * 60**2).to_i  unless value[:month].nil?
+    time = time + value[:day].to_i    * 24 * 60**2                   unless value[:day].nil?
+    time = time + value[:hour].to_i   * 60**2                        unless value[:hour].nil?
+    time = time + value[:minute].to_i * 60                           unless value[:minute].nil?
+    time = time + value[:second].to_i                                unless value[:second].nil?
+
+    time
+  end
+
+  # Returns the maximum length of time, in milliseconds, that the task
+  # will run before terminating.
+  #
+  # DONE
+  def max_run_time_as_ms
+    raise Error.new(_('No currently active task. ITask is NULL.')) if definition.nil?
+
+    # A value of PT0S will enable the task to run indefinitely.
+    max_time = time_limit_to_hash(definition.Settings.ExecutionTimeLimit)
+    return nil if max_time.nil?
+    
+    # # TODO Do I convert the code below into something portable?
+    # time = 0
+    # # Note - the Year and Month calculations are approximate
+    # time = time + max_time[:year].to_i   * (365.2422 * 24 * 60**2).to_i unless max_time[:year].nil?
+    # time = time + max_time[:month].to_i  * (365.2422 * 2 * 60**2).to_i  unless max_time[:month].nil?
+    # time = time + max_time[:day].to_i    * 24 * 60**2                   unless max_time[:day].nil?
+    # time = time + max_time[:hour].to_i   * 60**2                        unless max_time[:hour].nil?
+    # time = time + max_time[:minute].to_i * 60                           unless max_time[:minute].nil?
+    # time = time + max_time[:second].to_i                                unless max_time[:second].nil?
+
+    duration_hash_to_seconds(max_time) * 1000
+  end
+
+  # Sets the maximum length of time, in milliseconds, that the task can run
+  # before terminating. Returns the value you specified if successful.
+  #
+  def max_run_time=(max_run_time)
+    raise Error.new(_('No currently active task. ITask is NULL.')) if @pITask.nil?
+    raise TypeError unless max_run_time.is_a?(Numeric)
+
+    @pITask.SetMaxRunTime(max_run_time)
+    max_run_time
+  end
+
+  # Returns whether or not the scheduled task exists.
+  # DONE
+  def exists?(job_name)
+    # task name comparison is case insensitive
+    enum_task_names.any? { |name| name.casecmp(job_name) == 0 }
+  end
+
+  private
+
+  # Recursively converts a WIN32OLE Object in to a hash.  This method
+  # only outputs the Get Methods for an Object that has no parameters on the methods
+  # i.e. they are Object properties
+  #
+  def win32ole_to_hash(win32_obj)
+    hash = {}
+
+    win32_obj.ole_get_methods.each do |method|
```

Comment:
Need to answer the question around whether we want to do this dynamically like this... or just use the names of actual properties we care about.